### PR TITLE
Add randomized testing to test hundreds of different scenarios

### PR DESCRIPTION
**Pull request by labbati**
### Description

This PR adds randomized tests to our CI. Tests can also be run locally following instructions in `tests/randomized/README.md`.

This PR does not add the possibility to test for known regressions. This feature is to be added in a followup PR.

A few concepts to help the reviewer:
- no production code is touched in this PR, only tests
- The logical flow of running randomized tests and which files are relevant to that purpose are listed here:
  - you want to generate a bunch of sample applications: you use [`generate-scenarios.php`](https://github.com/DataDog/dd-trace-php/pull/1131/files#diff-883c8c9c53248d52887515f6a02b80e4cd1b0339e2d7b550155aa809c1518312).
    - Under the hood, it generates files into `.tmp.scenarios` that includes:
      - A general `Makefile` that can run all the tests: see `MakefileGenerator` class.
      - A `docker-compose.yml` file that will contain all the runners `DockerComposeFileGenerator` class.
      - Then each scenario in `.tmp.scenarios/<scenario-folder>` will have:
        - An apache configuration generator [`ApacheConfigGenerator.php`](https://github.com/DataDog/dd-trace-php/pull/1131/files#diff-220504bc310bd230f07b437bb1625fbfaa5e7132797af99b99b6eece7a386ab5)
        - A php-fpm configuration generator [`PhpFpmConfigGenerator.php`](https://github.com/DataDog/dd-trace-php/pull/1131/files#diff-cd46ab3ca4dce5fbfa1c0a41d7921c96f765fac5986cee2741a1283464ba3268)
        - A request target configuration generator [`RequestTargetGenerator.php`](https://github.com/DataDog/dd-trace-php/pull/1131/files#diff-f41cc4e5d3bfb20b0e9836f8a96033039247f3d3f6acb0bdf86e6e8c2374396c)
  - Analyze the results: see [`analyze-results.php`](https://github.com/DataDog/dd-trace-php/pull/1131/files#diff-28a4837595b4bca57b0799d3d7fd12a2cec420cc0ca47b4a424cedebc3a69da8)

Principles that are important:
- Fully reproducible. If I run `make generate SEED=123` in 2 different envs it should reproduce the same generated scenarios.
- During analysis, we should enforce that if there are less then N requests (for now is 1000) for one specific scenario, then the test should fail.

I suggest to download this branch go to `tests/randomized` and play with the commands in the [`README.md`](https://github.com/DataDog/dd-trace-php/pull/1131/files#diff-d55f111dd2c332d128569e9eac4c90e82ab38b1354ad961865ff13be50fd7eb2).

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

**Review by SammyK — COMMENTED** [review]
This is a really great start to the randomized testing framework. I think it will be a crucial tool for proactively detecting stability issues.

I'm assuming we'll want to upgrade and enhance this over time to be able to detect more edge cases and also update it as the tracer evolves. So it might make sense to make sure a few crucial pieces are more easily maintainable or extensible over time.

0. **Env + INI config:** These will probably change the most over time with env var deprecations, new env vars, migration to INI, etc. Currently it is not obvious that they are in the `tests/randomized/generate-scenarios.php` script. Perhaps these could have a separate documented file that would hold the config names and the possible values and/or ranges of values. Something like `tests/randomized/config/{env,ini}.php`?
0. **Templates:** I love the template config files.
    a. What are your thoughts on extracting the config generators into their own classes to make the config generation a bit more readable? Perhaps something akin to the [PHP-FPM template generator/runner](https://github.com/DataDog/dd-trace-php/blob/60b7423/tests/Sapi/PhpFpm/PhpFpm.php) that we use for tests.
    b. It might be beneficial to use variable substitution for the `docker-compose.template.yml` file generation (similar to the [template variable substitution for PHP-FPM](https://github.com/DataDog/dd-trace-php/blob/60b7423f914a49d6a30fd924b327b4cdadb3fede/tests/Sapi/PhpFpm/www.conf)). This would move the PHP service [out of PHP](https://github.com/DataDog/dd-trace-php/blob/19a3e4fa40b9919b98c09498b0693ce0e1d52027/tests/randomized/generate-scenarios.php#L211-L236) to make it easier to make changes going forward. For example, to add new docker containers for new integrations we would have to update the template file and then go looking for the `depends_on` key for the PHP service which is a bit buried in the `generate-scenarios.php` script.
0. **PHP versions:** We probably need to have a higher-level strategy for keeping track of PHP versions so that we don't miss something when adding new versions (but that's of course outside of the scope of this PR.) :) But perhaps we could extract the [array of OS/PHP versions](https://github.com/DataDog/dd-trace-php/blob/19a3e4fa40b9919b98c09498b0693ce0e1d52027/tests/randomized/generate-scenarios.php#L7-L21) into its own config file? Something like `tests/randomized/config/php_versions.php`?

For a future PR perhaps we could extract the random value generation to its own set of classes similar to the [Faker library](https://github.com/fzaninotto/Faker).

**Comment by labbati** [comment]
@SammyK thank you for the excellent review. @morrisonlevi I need to handle on second round of CR to you to move it forward with this project.

0. Excellent idea. Done
1.a and 1.b: Again, excellent idea. Done
2. as of now we only support cento7 and that installation method with those PHP versions, so they are very tight together. I agree that in the future we need to more flexible. I took a note so when we can discuss how to move forward, but I would not consider it to be a blocker for this release.

**Comment by morrisonlevi** [comment]
A smaller thing, but my only feedback so far:

Running locally, I see this classic warning in the shell:
> PHP Warning:  Unknown: It is not safe to rely on the system's timezone settings. You are *required* to use the date.timezone setting or the date_default_timezone_set() function. In case you used any of those methods and you are still getting this warning, you most likely misspelled the timezone identifier. We selected the timezone 'UTC' for now, but please set date.timezone to select your timezone. in Unknown on line 0

It's unclear to me which exact container/PHP version was emitting this as builds were going in parallel, but maybe 5.5? We should fix it in the container, I think. The base php.ini should set a timezone to `UTC`, which is what the default is in newer PHP versions (I think).

**Comment by morrisonlevi** [comment]
I feel like this is downloading WAY too many things relative to the amount of testing being done. I did a `make generate NUMBER_OF_SCENARIOS=1337` and it's spending near 100% of the time downloading and installing stuff.

Also, I'm not sure how to interpret these results:

```
Analyzed 31 scenarios.
Minimum request not matched: array (
  'randomized-1155509609-centos7-5.4' => 389,
  'randomized-1161705451-centos7-7.3' => 239,
  'randomized-1309666788-centos7-7.1' => 165,
  'randomized-1379213473-centos7-7.3' => 231,
  'randomized-1383060589-centos7-7.4' => 165,
  'randomized-1549733249-centos7-7.3' => 659,
  'randomized-1637771703-centos7-8.0' => 347,
  'randomized-1745595390-centos7-5.5' => 328,
  'randomized-1753096970-centos7-7.0' => 164,
  'randomized-1785295490-centos7-5.6' => 135,
  'randomized-1884356983-centos7-5.6' => 170,
  'randomized-1884959314-centos7-7.0' => 322,
  'randomized-190171776-centos7-8.0' => 497,
  'randomized-1910460837-centos7-8.0' => 232,
  'randomized-1931037473-centos7-7.0' => 382,
  'randomized-1968649742-centos7-5.6' => 420,
  'randomized-2087550935-centos7-5.5' => 168,
  'randomized-2098369243-centos7-7.3' => 176,
  'randomized-245380129-centos7-7.4' => 322,
  'randomized-293302327-centos7-5.5' => 384,
  'randomized-294241521-centos7-7.3' => 188,
  'randomized-307537664-centos7-7.1' => 129,
  'randomized-372196224-centos7-5.4' => 945,
  'randomized-430456162-centos7-5.5' => 490,
  'randomized-452355447-centos7-5.5' => 335,
  'randomized-560870433-centos7-5.6' => 274,
  'randomized-614902931-centos7-5.4' => 212,
  'randomized-774519797-centos7-7.2' => 245,
  'randomized-841749501-centos7-5.6' => 176,
  'randomized-906873934-centos7-8.0' => 325,
  'randomized-959248483-centos7-5.5' => 334,
)
make: *** [analyze] Error 1
```

**Comment by labbati** [comment]
Thank you @morrisonlevi for the precious code review. Let me address you four points separately.

> PHP Warning: Unknown: It is not safe to rely on the system's timezone settings. You are required to use the date.timezone setting or the date_default_timezone_set() function. In case you used any of those methods and you are still getting this warning, you most likely misspelled the timezone identifier. We selected the timezone 'UTC' for now, but please set date.timezone to select your timezone. in Unknown on line 0

Good catch. This is fixed by commit https://github.com/DataDog/dd-trace-php/pull/1131/commits/0858a22e3e73ce6b4a1efcefd70bee15be10b43e.

> I feel like this is downloading WAY too many things relative to the amount of testing being done. I did a make generate NUMBER_OF_SCENARIOS=1337 and it's spending near 100% of the time downloading and installing stuff. I think I've downloaded Guzzle 6.5.5 quite a few times.

Actually it fools you. The last log message mentions `guzzle`, but actually the composer library consuming most of the time it's...us. An investigation and mitigation is in [APMPHP-535](https://datadoghq.atlassian.net/browse/APMPHP-535). I don't envision addressing this as part of this PR but as a separate optimization/followup.

Also, I cannot share a composer cache folder as it generates lock/wait errors that make the test too flaky. Once we fix the size of our package, though, it will be a matter of ~10 seconds only, which is totally acceptable.

> Edit: I eventually filled the disk space for containers. Maybe I'm just using it wrong?

This is docker and I was partially able to mitigate this issue with https://github.com/DataDog/dd-trace-php/pull/1131/commits/89c0da59d29e9156454b0fa5493549472e25e22b.

Before: PHP and other binaries installation took `349MB`.

```
$ docker history -H datadog/dd-trace-ci:php-randomizedtests-centos7-7.4
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
4bee42706bae   19 minutes ago   CMD ["bash" "/scripts/run.sh"]                  0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   ENV COMPOSER_VENDOR_DIR=/composer-vendor        0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   ENV COMPOSER_CACHE_DIR=/composer-cache          0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   WORKDIR /var/www/html                           0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   ADD prepare.sh /scripts/prepare.sh # buildkit   624B      buildkit.dockerfile.v0
<missing>      19 minutes ago   ADD run.sh /scripts/run.sh # buildkit           324B      buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ec…   11.8kB    buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c se…   11.8kB    buildkit.dockerfile.v0
<missing>      19 minutes ago   ADD nginx.conf /etc/nginx/conf.d/default.con…   530B      buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ad…   359kB     buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c gr…   1.45kB    buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c se…   19.5kB    buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      19 minutes ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ec…   20B       buildkit.dockerfile.v0
<missing>      19 minutes ago   ADD enable-coredump.sh /scripts/enable-cored…   256B      buildkit.dockerfile.v0
<missing>      2 weeks ago      RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ch…   0B        buildkit.dockerfile.v0
<missing>      2 weeks ago      RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      2 weeks ago      RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c cu…   2.22MB    buildkit.dockerfile.v0
<missing>      2 weeks ago      RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ta…   8.69MB    buildkit.dockerfile.v0
<missing>      2 weeks ago      ADD https://github.com/tsenart/vegeta/releas…   3.55MB    buildkit.dockerfile.v0
<missing>      2 weeks ago      RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c rp…   349MB     buildkit.dockerfile.v0
<missing>      2 weeks ago      RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ec…   109B      buildkit.dockerfile.v0
<missing>      2 weeks ago      ARG PHP_MINOR                                   0B        buildkit.dockerfile.v0
<missing>      2 weeks ago      ARG PHP_MAJOR                                   0B        buildkit.dockerfile.v0
<missing>      3 months ago     /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B
<missing>      3 months ago     /bin/sh -c #(nop)  LABEL org.label-schema.sc…   0B
<missing>      3 months ago     /bin/sh -c #(nop) ADD file:b3ebbe8bd304723d4…   204MB
```

After: PHP and other binaries installation took `226MB`, which is `125MB` less. Considering that `centos:7`'s `204MB` are very likely to be cached by their system, I am optimistic that this is an acceptable tradeoff. I do not have other optimizations I can think of right now, but I am open to ideas, of course.

```
$ docker history -H datadog/dd-trace-ci:php-randomizedtests-centos7-7.4
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
8524a16008dc   About a minute ago   CMD ["bash" "/scripts/run.sh"]                  0B        buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      About a minute ago   ENV COMPOSER_VENDOR_DIR=/composer-vendor        0B        buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      About a minute ago   ENV COMPOSER_CACHE_DIR=/composer-cache          0B        buildkit.dockerfile.v0
<missing>      About a minute ago   WORKDIR /var/www/html                           0B        buildkit.dockerfile.v0
<missing>      About a minute ago   ADD prepare.sh /scripts/prepare.sh # buildkit   624B      buildkit.dockerfile.v0
<missing>      About a minute ago   ADD run.sh /scripts/run.sh # buildkit           324B      buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ec…   11.8kB    buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c se…   11.8kB    buildkit.dockerfile.v0
<missing>      About a minute ago   ADD nginx.conf /etc/nginx/conf.d/default.con…   530B      buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ad…   359kB     buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c gr…   1.45kB    buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c se…   19.5kB    buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ec…   20B       buildkit.dockerfile.v0
<missing>      About a minute ago   ADD enable-coredump.sh /scripts/enable-cored…   256B      buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ch…   0B        buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c mk…   0B        buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c cu…   2.23MB    buildkit.dockerfile.v0
<missing>      About a minute ago   RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c cu…   8.71MB    buildkit.dockerfile.v0
<missing>      7 minutes ago        RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c rp…   226MB     buildkit.dockerfile.v0
<missing>      2 weeks ago          RUN |2 PHP_MAJOR=7 PHP_MINOR=4 /bin/sh -c ec…   109B      buildkit.dockerfile.v0
<missing>      2 weeks ago          ARG PHP_MINOR                                   0B        buildkit.dockerfile.v0
<missing>      2 weeks ago          ARG PHP_MAJOR                                   0B        buildkit.dockerfile.v0
<missing>      3 months ago         /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B
<missing>      3 months ago         /bin/sh -c #(nop)  LABEL org.label-schema.sc…   0B
<missing>      3 months ago         /bin/sh -c #(nop) ADD file:b3ebbe8bd304723d4…   204MB
```

> Also, I'm not sure how to interpret these results from a smaller run:

Long story short, I want to make sure that at least a minimum number of requests are executed against any specific scenario. I added this explanation, I hope it helps clarifying https://github.com/DataDog/dd-trace-php/pull/1131/commits/c1f8be881e4d30910af0fb78dc001e1527caffef.

File: tests/randomized/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+Currently the following checks are executed:
+- the returned status code can only be on of 200 (OK), 510 (controlled uncaught exception), 511 (controlled php error). Any other return code will result in a failing analysis.
+- At least 1000 requests have been executed. If we find that vegeta has performed less than 1000 request, then the  test fails as we might have put the system not under enough pressure and combination of execution paths to find meaningful problems. In this case the error will be reported with a message `Minimum request not matched` and, unless there are not requests at all, it typically means that your hardward is not power enough to run a large amount of concurrent tests. You can either reduce concurrency via `make execute CONCURRENT_JOBS=3` or run specific tests via `make -C .tmp.scenarios test.scenario.<scenario_name>`.
+
+## Debugging a segmentation fault
+
+At the beginning of every request a message is printed to correlate a PID and a seed. E.g. `Current PID: 70. Current seed 754539563`.
+That seed can be used later on to recreate the same request: `curl localhost:80?seed=754539563`.
```
Comment (morrisonlevi) [inline tests/randomized/README.md:101]: How do I bring up the local web server?

Comment (labbati) [reply tests/randomized/README.md:126]: Very good point @morrisonlevi. I added these few lines to the README.md, hope this helps: https://github.com/DataDog/dd-trace-php/pull/1131/commits/de2aa8f23b29b74cc4ee9c94307498249e4113ca

Comment (morrisonlevi) [reply tests/randomized/README.md:126]: I will try this today. Being able to go from "found an error" to "having a backtrace in my gdb" is an important piece for this tool to be useful, so it's the last part of my review.

Comment (labbati) [reply tests/randomized/README.md:126]: Good point. Also, I decided to backport https://github.com/DataDog/dd-trace-php/pull/1131/commits/88440293086c824e724cd4e0ee291efdb079b483 from the branch providing regression freezing. This will increase reproducibility.

**Comment by labbati** [comment]
@morrisonlevi I pushed this fix https://github.com/DataDog/dd-trace-php/commit/af2d141a970845621e27166d1a57c937893004b4 to reduce flakiness. Explanation is in the code comment

**Review by morrisonlevi — APPROVED** [review]
Approved. My only concern going forward is maintaining lists of integrations, environment variables, etc.

**Comment by labbati** [comment]
@morrisonlevi thanks for the CR. Your concerns is valid, let's see how mitigate that or improve how we build the randomized app as we add integrations and configurations
